src/server/lock_manager/deadlock.rs

```diff
@@ -305,32 +407,177 @@ impl DetectTable {
             self.now,
         );
         wait_for.insert(locks.ts, locks);
+
+        Self::insert_key_reverse_index(&mut self.key_reverse_map, key, lock_ts, txn_ts);
     }
 
     /// Removes the corresponding wait_for_entry.
-    fn clean_up_wait_for(&mut self, txn_ts: TimeStamp, lock_digest: LockDigest) {
+    fn clean_up_wait_for(&mut self, txn_ts: TimeStamp, lock_digest: LockDigest, key: &[u8]) {
         if let Some(wait_for) = self.wait_for_map.get_mut(&txn_ts) {
             if let Some(locks) = wait_for.get_mut(&lock_digest.ts) {
-                if locks.remove(lock_digest.hash) {
+                let removed_lock = locks.remove(lock_digest.hash, key);
+                if locks.is_empty() {
                     wait_for.remove(&lock_digest.ts);
                     if wait_for.is_empty() {
                         self.wait_for_map.remove(&txn_ts);
                     }
                 }
+
+                // If the lock-waiting is registered by a TiKV node in new version, but
+                // the `clean_up_wait_for` is sent from a node in old version (before v5.1), the
+                // parameter `key` can be empty while the `wait_for_map` has stored the key.
+                // In this case, use the key from the deleted entry from `locks`.
+                if let Some((_, removed_key)) = removed_lock {
+                    Self::remove_key_reverse_index(
+                        &mut self.key_reverse_map,
+                        &removed_key,
+                        lock_digest.ts,
+                        txn_ts,
+                    )
+                }
             }
         }
+        self.check_key_reverse_index_consistency();
         TASK_COUNTER_METRICS.clean_up_wait_for.inc();
     }
 
+    /// Replaces the specific lock on the specific key, and re-detect all
+    /// waiters on this lock. For each detected deadlocks, returns a tuple
+    /// `(txn_id, key_hash, key, wait_chain, resource_group_tag)`. Note that
+    /// the `key_hash` and `key` differs from the one given in parameter. Same
+    /// as the `detect` function's return value, the key is the one on which
+    /// the transaction aborting by the deadlock blocks another transaction
+    /// in the deadlock circle.
+    fn replace_lock_by_key(
+        &mut self,
+        lock_hash: u64,
+        key: Vec<u8>,
+        old_lock_ts: TimeStamp,
+        new_lock_ts: TimeStamp,
+    ) -> Vec<(TimeStamp, u64, Vec<u8>, Vec<WaitForEntry>, Vec<u8>)> {
+        if old_lock_ts == new_lock_ts {
+            return vec![];
```

Comment:
This is path is here to make the function's behavior as much well-defined as possible. If we consider it deeply, this path is actually theoretically possible. Consider several pessimistic-lock requests in non-fair-locking mode is waiting in queue, and a transaction A holds the lock on this key; then, transaction A retries the current statement, so that it pessimisitc-rollback the lock and then tries to acquire the lock again. if transaction A's RPC is fast enough, it's theoretically possible that A successfully acquires the lock before any other woken up waiters or other concurrent transactions. So for these requests in the waiting queue, they may observe that the lock of A is release and another lock of A is then acquired.